### PR TITLE
Feature/token holders

### PR DESCRIPTION
**Pull request by OgnjenCBS**

### CODE REVIEW

File: src/modules/token-holders/token-holders.service.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    const allHolders = await this.holderRepo.findAllHolders();
+    allHolders.forEach((holder) => {
+      if (holder.balance === 0) {
+        this.holderRepo.deleteHolder(holder);
+      }
+    });
+    this.logger.log('Holders with 0 balance deleteds');
+  }
+
+  public async getHoldersAndBalances(
```
Comment (DanijelCBS) [inline src/modules/token-holders/token-holders.service.ts:109]: Move this method to the top.

File: src/modules/token-holders/token-holders.service.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    const holders = await this.holderRepo.findHoldersAndBalances(
+      pageOptions,
+      token,
+    );
+
+    return holders;
+  }
+
+  @Cron(CronExpression.EVERY_30_MINUTES)
+  private async updateHolders(): Promise<void> {
```
Comment (DanijelCBS) [inline src/modules/token-holders/token-holders.service.ts:122]: Move this method to be the 2nd in the file.

File: src/modules/token-holders/token-holders.service.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+          holderEntity.balance = asset.balance;
+          this.holderRepo.upsertHolder(holderEntity);
+        });
+      }),
+    );
+
+    this.logger.log('Updating holders balances successful.');
+  }
+
+  private async removeZeroBalances(): Promise<void> {
```
Comment (DanijelCBS) [inline src/modules/token-holders/token-holders.service.ts:98]: Check if this can be done in the repo with delete method and where condition.

**Review by DanijelCBS — CHANGES_REQUESTED** [review]

**Review by DanijelCBS — APPROVED** [review]

File: src/database/mysql/migrations/1703504742036-create-holders-token-balances.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+export class createHoldersTokenBalances1703504742036
+  implements MigrationInterface
+{
+  name = 'createHoldersTokenBalances1703504742036';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`CREATE TABLE \`holders\` (
+            \`id\` int NOT NULL AUTO_INCREMENT, 
+            \`holder\` varchar(255) NOT NULL,
```
Comment (Str4LeGit) [inline src/database/mysql/migrations/1703504742036-create-holders-token-balances.ts:11]: Is this address format? We should limit the length to 40 or 50, 255 is not necessary.

File: src/database/mysql/migrations/1703504742036-create-holders-token-balances.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+export class createHoldersTokenBalances1703504742036
+  implements MigrationInterface
+{
+  name = 'createHoldersTokenBalances1703504742036';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`CREATE TABLE \`holders\` (
+            \`id\` int NOT NULL AUTO_INCREMENT, 
+            \`holder\` varchar(255) NOT NULL, 
+            \`token\` varchar(255) NOT NULL,
```
Comment (Str4LeGit) [inline src/database/mysql/migrations/1703504742036-create-holders-token-balances.ts:12]: Same here.

File: src/modules/token-holders/token-holders.service.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        }[] = [
+          ...portfolio.map((portfolioAsset) => {
+            const [assetsId, assetAmount] = portfolioAsset;
+            const [, { code: assetId }] = assetsId.toHuman();
+            const { free: assetBalance } = assetAmount.toHuman();
+
+            return {
+              token: assetId,
+              balance: new FPNumber(assetBalance)
+                .div(FPNumber.fromNatural(10 ** 18))
```
Comment (Str4LeGit) [inline src/modules/token-holders/token-holders.service.ts:98]: There should be denominator constant that can be reused.

**Review by Str4LeGit — APPROVED** [review]

**Pull request merged**